src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -43,6 +45,153 @@ void Bootstrap(const MatType& dataset,
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    bootstrapDataset = dataset;
+    bootstrapLabels = labels;
+    if (UseWeights)
+      bootstrapWeights = weights;
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ */
+template<bool UseWeights, typename G = std::mt19937>
+class SequentialBootstrap
+{
+ public:
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
+                               const G& generator = G(std::random_device()())) :
+    SequentialBootstrap(indicatorMatrix, indicatorMatrix.n_rows, generator)
+  {
+  }
+
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
+                               arma::uword sampleCount,
+                               const G& generator = G(std::random_device()())):
+    indicatorMatrix(indicatorMatrix),
+    colIndices(arma::linspace<arma::uvec>(0, indicatorMatrix.n_cols - 1)),
+    sampleCount(sampleCount),
+    generator(generator)
+  {
+  }
+
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    if (labels.n_cols != indicatorMatrix.n_rows)
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+          "labels and indicatorMatrix n_rows differ!");
+
+    // observations are stored as columns and dimensions
+    // (number of features) as rows.
+
+    const arma::uvec phi(ComputeSamples());
+
+    bootstrapDataset = dataset.cols(phi);
+    bootstrapLabels = labels.cols(phi);
+    if (UseWeights)
+      bootstrapWeights = weights.rows(phi);
+  }
+
+  arma::uvec ComputeSamples() const
+  {
+    arma::uvec phi;
+
+    while (phi.size() < sampleCount) {
+      const auto prob(ComputeNextDrawProbabilities(phi, indicatorMatrix));
+      std::discrete_distribution dist(prob.begin(), prob.end());
+
+      phi = arma::join_cols(phi, arma::uvec{ colIndices[dist(generator)] });
+    }
+
+    return phi;
+  }
+
+  static arma::vec ComputeAverageUniqueness(const arma::mat& indicatorMatrix)
+  {
+    // sum of each column
+    arma::rowvec concurrency(arma::sum(indicatorMatrix, 0));
+    auto uniqueness(
+      indicatorMatrix.each_row()
+      / concurrency.clamp(1.0,
+        std::numeric_limits<arma::rowvec::elem_type>::max()));
+    const auto n(arma::sum(indicatorMatrix, 1)); // sum of each row
+
+    return arma::sum(uniqueness, 1) / n; // mean of each row
+  }
+
+  static arma::vec ComputeNextDrawProbabilities(const arma::uvec& phi,
+                                             const arma::mat& indicatorMatrix)
```

Comment:
This function and `ComputeAverageUniqueness()` (and actually `ComputeSamples()` too) are simple enough and only used once, so I would actually just inline them.  That also helps avoid the issue of copying around `arma::vec`s from return values.  (I don't generally trust compilers to get that right.)

Reply:
I explicitly created those functions to test them in unit tests.